maven/JarAssembler.kt

```diff
@@ -34,35 +36,80 @@ class JarAssembler : Callable<Unit> {
+        }
+        return newEntries
+    }
+
+    private fun getFinalPath(entry: ZipEntry, sourceFileBytes: ByteArray): String {
+        return if (entry.name.endsWith(".java")) {
+            // files in source JARs are moved according to their `package` statement
+            val sourceFile = toStringUTF8(sourceFileBytes)
+            val sourceFileName = Paths.get(entry.name).fileName.toString()
+            val sourceFilePackage = getJavaPackage(sourceFile) ?: throw RuntimeException("could not obtain package of ${entry.name}")
+            "${sourceFilePackage.replace(".", "/")}/$sourceFileName"
+        } else {
+            entry.name
+        }
+    }
+
+    private fun toStringUTF8(sourceFileBytes: ByteArray): String {
+        return sourceFileBytes.toString(Charset.forName("UTF-8"))
+    }
+
+    private fun getJavaPackage(sourceFile: String): String? {
+        val javaPackageRegex = Regex("package\\s+([a-zA_Z_][\\.\\w]*);")
+        return javaPackageRegex.find(sourceFile)?.groups?.get(1)?.value
+    }
 
     override fun call() {
         ZipOutputStream(BufferedOutputStream(FileOutputStream(output_file))).use { out ->
             if (pomFile != null) {
-                val pomFileEntry = ZipEntry("META-INF/maven/${groupId}/${artifactId}/pom.xml")
-                out.putNextEntry(pomFileEntry)
-                out.write(pomFile!!.readBytes())
+                val pomPath = "META-INF/maven/${groupId}/${artifactId}/pom.xml"
+                entries += preCreateDirectories(Paths.get(pomPath))
+                entries[pomPath] = pomFile!!.readBytes()
             }
             for (jar in jars) {
                 ZipFile(jar).use { jarZip ->
                     jarZip.entries().asSequence().forEach { entry ->
                         if (entry.name.contains("META-INF")) {
+                            // pom.xml will be added by us
                             return@forEach
                         }
                         if (entryNames.contains(entry.name)) {
+                            throw RuntimeException("duplicate entry in the JAR: ${entry.name}")
+                        }
+                        if (entry.isDirectory) {
+                            // needed directories would be added by us
                             return@forEach
                         }
                         entryNames.add(entry.name)
                         BufferedInputStream(jarZip.getInputStream(entry)).use { inputStream ->
-                            val newEntry = ZipEntry(prefix + entry.name)
-                            out.putNextEntry(newEntry)
-                            inputStream.copyTo(out, 1024)
+                            val sourceFileBytes = inputStream.readBytes()
+                            val resultLocation = getFinalPath(entry, sourceFileBytes)
+                            entries[resultLocation] = sourceFileBytes
+                            preCreateDirectories(Paths.get(resultLocation))
```

Comment:
Are we sure we don't need to do `entries += preCreateDirectories(Paths.get(resultLocation))`?

Reply:
@lolski good point; addressed in https://github.com/vaticle/bazel-distribution/pull/305/commits/ea631b9c791ec6fc11151abc9f58d5f9795408d5